### PR TITLE
Implement IP filter

### PR DESCRIPTION
**Pull request by orion0616**
Implement IP filter
- enable to filter IP address by using Trie

### CODE REVIEW

File: lib/fluent/plugin/filter_filter_list.rb [hunk 1]
```diff
@@ -1,21 +1,24 @@
...
 
 module Fluent
   module Plugin
     class FilterListFilter < Filter
       include Matchers
+      include IP
 
       Plugin.register_filter('filter_list', self)
 
+      config_param :filter, :string, default: 'AC'
```
Comment (yanana) [inline lib/fluent/plugin/filter_filter_list.rb:14]: Please describe the new configuration on README.

File: lib/fluent/plugin/filter_filter_list.rb [hunk 2]
```diff
@@ -1,21 +1,24 @@
...
       config_param :key_to_filter, :string, default: nil
       config_param :patterns_file_path, :string, default: ''
 
       def configure(conf)
         super
         patterns = @patterns_file_path.empty? ? [] : File.readlines(@patterns_file_path).map(&:chomp).reject(&:empty?)
-        @matcher = ACMatcher.new(patterns)
+        @matcher = (@filter == 'AC') ? ACMatcher.new(patterns) : IPMatcher.new(patterns)
```
Comment (yanana) [inline lib/fluent/plugin/filter_filter_list.rb]: I'd like the default matcher to be the _AC_. So when `@filter` is unknown, please make `@matcher` an instance of `ACMatcher`.

File: lib/matcher.rb [hunk 3]
```diff
@@ -62,6 +81,17 @@ def build
         end
       end
     end
+
+    def foward_match(pattern)
```
Comment (yanana) [inline lib/matcher.rb]: `s/foward/forward/g`

File: test/fluent/plugin/filter_filter_list_test.rb [hunk 4]
```diff
@@ -49,6 +55,18 @@ def test_that_message_containing_a_pattern_is_filtered
         assert_equal 'ab', es[0]['x']
         assert_equal 'foo', es[0]['y']
       end
+
+      def test_that_message_containing_an_ip_is_filtered
+        d = create_driver(CONFIG2)
+        d.run(default_tag: 'test') do
+          d.feed('ip' => '192.168.1.1', 'y' => 'foo')
```
Comment (yanana) [inline test/fluent/plugin/filter_filter_list_test.rb:62]: Please provide a little more boundary cases.

File: lib/fluent/plugin/filter_filter_list.rb (hunk 1 above)
Comment (yanana) [reply lib/fluent/plugin/filter_filter_list.rb:14]: The format of the list of IP addresses also should be described. Especially, the list accepts both usual IP address and a CIDR-notated group of IP addresses as an entry.

File: lib/ip.rb [hunk 5]
```diff
@@ -0,0 +1,13 @@
+module IP
+  require 'ipaddr'
+  class IP
+    def initialize(ip_with_cidr)
```
Comment (yanana) [inline lib/ip.rb]: The constructor accepts not only a CIDR-notated IP but also just an exact IP address. So please comment it, since I couldn't understand the behavior from the argument's name.

File: lib/fluent/plugin/filter_filter_list.rb (hunk 2 above)
Comment (orion46) [reply lib/fluent/plugin/filter_filter_list.rb]: I think `@filter`'s default value must be `AC` in line 14.

**Review by yanana — APPROVED** [review]
LGTM! Thanks 😄

**Review by yanana — COMMENTED** [review]
Could you please review the cause of Travis failure?

File: lib/ip.rb [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+  require 'ipaddr'
+  class IP
+    # This constructor accepts both CIDR-notated IP and also exact IP.
+    def initialize(ip)
+      @address = IPAddr.new(ip)
+    end
+
+    def to_binary
+      binary = @address.to_i.to_s(2)
+      binary.slice(0, @address.prefix)
```
Comment (yanana) [inline lib/ip.rb]: Since `IPAddr#prefix` was introduced in Ruby 2.5.0, you have to provide an alternative way to acquire the prefix length of IP addresses. Unless `prefix` is defined, define the method on the class (monkey-patching).
